From b6df9bea10f870c8fe87328b200cabfc117b5372 Mon Sep 17 00:00:00 2001
From: Mika Laitio <mika.laitio@amd.com>
Date: Wed, 6 Aug 2025 01:58:52 -0700
Subject: [PATCH] Keep only one plus-character in version suffix

Triton version can only contain one plus-character. The plus-character
is used for distinguishing additional version suffix information from
the base major.minor.patch version scheme.

If the Triton build is made from the release branch, then
get_git_version_suffix() will return "". In every other case it will return
"+githash". (One plus character in the beginning as the githash itself
does not contain plus-characters)

If the build is done from the release-branch and TRITON_WHEEL_VERSION_SUFFIX
environment variable is used, then the "+githash" is not added to version suffix
and env[TRITON_WHEEL_VERSION_SUFFIX] should start with the plus-character.
In other cases like nightly builds the "+githash" is added to version suffix
and then env[TRITON_WHEEL_VERSION_SUFFIX] should not contain plus-character.

This kind of checking adds extra complexity to the build script side and easier
solution is to check in setup.py whether the env[TRITON_WHEEL_VERSION_SUFFIX]
should contain the plus-character or not and if needed replace the
plus-character with the dash-character.

Example from the error fixed by this patch:

File ".venv_sdk/lib/python3.12/site-packages/packaging/version.py", line 202, in __init__
    raise InvalidVersion(f"Invalid version: {version!r}")
    packaging.version.InvalidVersion: Invalid version: '3.4.0+gitd04b0f4a+rocmsdk20250806'

Co-authored-by: Marius Brehler <marius.brehler@amd.com>
Signed-off-by: Mika Laitio <mika.laitio@amd.com>
---
 setup.py | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index be40c6eb..f479e90f 100644
--- a/setup.py
+++ b/setup.py
@@ -771,8 +771,19 @@ def get_git_version_suffix():
         return get_git_commit_hash()
 
 
+def get_triton_version_suffix():
+    # Either "" or "+<githash>", "<githash>" itself does not contain any plus-characters.
+    git_sfx = get_git_version_suffix()
+    # Should start with "+" that will replaced with "-" if needed
+    env_sfx = os.environ.get("TRITON_WHEEL_VERSION_SUFFIX", "")
+    # version suffix can only contain one plus-character
+    if "+" in git_sfx and "+" in env_sfx:
+        env_sfx = env_sfx.replace("+", "-")
+    return git_sfx + env_sfx
+
+
 # keep it separate for easy substitution
-TRITON_VERSION = "3.4.0" + get_git_version_suffix() + os.environ.get("TRITON_WHEEL_VERSION_SUFFIX", "")
+TRITON_VERSION = "3.4.0" + get_triton_version_suffix()
 
 # Dynamically define supported Python versions and classifiers
 MIN_PYTHON = (3, 9)
-- 
2.43.0

